src/server.ts

```diff
@@ -212,234 +223,256 @@ export class SubscriptionServer {
+      } else {
+        let executor: SubscribeFunction | ExecuteFunction = this.execute;
+        if (this.subscribe && isASubscriptionOperation(document, params.operationName)) {
+          executor = this.subscribe;
+        }
+        executionPromise = Promise.resolve(
+          executor(params.schema, document, this.rootValue, params.context, params.variables, params.operationName),
+        );
+      }
+      const executionResult = await executionPromise;
+      const executionIterable = isAsyncIterable(executionResult) ? executionResult : createAsyncIterator([executionResult]);
+      if (operationContext.executionIterator) {
+        this.unsubscribe(connectionContext, opId);
+      }
+      operationContext.executionIterator = executionIterable as ExecutionIterator;
+      // NOTE: This is a temporary code to support the legacy protocol.
+      // As soon as the old protocol has been removed, this code should also be removed.
+      this.sendMessage(connectionContext, opId, MessageTypes.SUBSCRIPTION_SUCCESS, undefined);
+      operationContext.processingStart = false;
+      nextTick(() => {
+        this.processQueue(connectionContext, opId);
+      });
+      await forAwaitEach(executionIterable as any, (value: ExecutionResult) => {
+        let result = value;
+
+        if (params.formatResponse) {
+          try {
+            result = params.formatResponse(value, params);
+          } catch (err) {
+            console.error('Error in formatError function:', err);
+          }
+        }
+        this.sendMessage(connectionContext, opId, MessageTypes.GQL_DATA, result);
+      });
+      this.sendMessage(connectionContext, opId, MessageTypes.GQL_COMPLETE, null);
+      nextTick(() => {
+        this.processQueue(connectionContext, opId);
+      });
+    } catch (error) {
+      operationContext.processingStart = false;
+      nextTick(() => {
+        this.processQueue(connectionContext, opId);
+      });
+      if (error.errors) {
+        this.sendMessage(connectionContext, opId, MessageTypes.GQL_DATA, {
+          errors: error.errors,
+        });
+      } else {
+        this.sendError(connectionContext, opId, { message: error.message });
+      }
+      this.unsubscribe(connectionContext, opId);
+    }
+  }
+
+  private processQueue(connectionContext: ConnectionContext, opId: string) {
+    const operationContext = connectionContext.operations.get(opId);
+    if (!operationContext || operationContext.processingStart) {
+      return;
+    }
+    const message = operationContext.messageQueue.shift();
+    if (!message) {
+      return;
+    }
+    const { type } = message;
+    if (type === MessageTypes.GQL_STOP) {
+      nextTick(() => {
+        this.processQueue(connectionContext, opId);
+      });
+      this.unsubscribe(connectionContext, opId);
+      return;
+    }
+    // START
+    this.processStart(connectionContext, operationContext, opId, message).catch(error => {
```

Comment:
Would probably be good to be defensive here and check the type in case of an accidental code change to push a different message type onto the queue.